src/ioabc.cpp

```diff
@@ -99,143 +99,66 @@ bool ABCInput::Import(const std::string &abc)
 // parseABC --
 //
 
-void ABCInput::parseABC(std::istream &infile)
+void ABCInput::ParseABC(std::istream &infile)
 {
     // initialize doc
     m_doc->Reset();
     m_doc->SetType(Raw);
 
-    // read file header
-    std::getline(infile, abcLine);
-    while (abcLine[0] != 'X') {
-        if (infile.eof()) {
-            LogError("ABC import: No tune found");
-            exit(0);
-        }
-        if ((abcLine[0] == '%') && (abcLine[1] == '%'))
-            LogWarning("ABC import: Stylesheet directives are ignored");
-        else if (abcLine[1] == ':')
-            readInformationField(abcLine[0], &abcLine[2]);
-        std::getline(infile, abcLine);
-        ++m_lineNum;
-    }
+    Score *score = NULL;
+    Section *section = NULL;
     CreateHeader();
-
     while (!infile.eof()) {
-        while (!(abcLine[0] == 'X' && abcLine[1] == ':') && !infile.eof()) {
-            std::getline(infile, abcLine);
-            ++m_lineNum;
-        }
-        if (infile.eof()) break;
-
-        // read tune header
-        readInformationField('X', &abcLine[2]);
-        while (abcLine[0] != 'K' && !infile.eof()) {
-            std::getline(infile, abcLine);
-            ++m_lineNum;
-            readInformationField(abcLine[0], &abcLine[2]);
-        }
-        if (infile.eof()) break;
-        if (m_title.empty()) {
-            LogWarning("ABC import: Title field missing, creating empty title");
-            m_title.push_back(std::make_pair("", 0));
-        }
-        // add work entry to meiHead
-        CreateWorkEntry();
-
-        // create score
-        assert(m_mdiv != NULL);
-        Score *score = new Score();
-        if (!m_doc->HasCurrentScore() || !m_doc->GetCurrentScoreDef()->GetFirst(STAFFGRP)) {
-            m_mdiv->AddChild(score);
-
-            // create page head
-            PrintInformationFields();
-            StaffGrp *staffGrp = new StaffGrp();
-            // create staff
-            StaffDef *staffDef = new StaffDef();
-            staffDef->SetN(1);
-            staffDef->SetLines(m_stafflines);
-            staffDef->SetTransSemi(m_transpose);
-            if (m_clef) {
-                staffDef->AddChild(m_clef);
-                m_clef = NULL;
-            }
-            if (m_meter) {
-                staffDef->AddChild(m_meter);
-                m_meter = NULL;
-            }
-            staffGrp->AddChild(staffDef);
-            m_doc->GetCurrentScoreDef()->AddChild(staffGrp);
-            if (m_key) {
-                m_doc->GetCurrentScoreDef()->AddChild(m_key);
-                m_key = NULL;
+        std::getline(infile, abcLine);
+        ++m_lineNum;
+        if (!abcLine.empty() && (abcLine.at(0) == 'X')) {
+            while (!infile.eof()) {
+                if (abcLine.length() >= 3) {
+                    this->readInformationField(abcLine.at(0), &abcLine.at(2));
```

Comment:
A std::string's data buffer is not required to be null terminated.  Must use c_str() to guarantee a null terminator.  So instread of  &abcLine.at(2) suggest either
&abcLine.c_str()[2]
or
abcLine.substr(2)

There are other similar usages in this file that need to be fixed.

Reply:
using `.substr()` looks good.